File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public HashMap<Character, Character> pair;

    public int solution(String s) {
        int answer = 0;
        pair = new HashMap<>();
        pair.put(')', '(');
        pair.put('}', '{');
        pair.put(']', '[');

        
        String rotated = s;
        for (int i = 0; i < s.length(); i++) {
            if(isValid(rotated)) answer++;
            
            rotated = rotated.substring(1) + rotated.charAt(0);

        }
        
        
        return answer;
    }
    
    private boolean isValid(String s){
        Stack<Character> check = new Stack<>();
        
        for(char ch: s.toCharArray()) {
            if(!pair.containsKey(ch)){
                check.push(ch);
                continue;
            }

            if(check.isEmpty() || pair.get(ch)!=check.pop()) return false;


        }
        return check.isEmpty();
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def right_pairs(parentheses):
    if len(parentheses) % 2 != 0: return False

    stack = []
    pairs = {'}':'{', ')':'(', ']':'['}
    
    for p in parentheses:
        if p in '{([': 
            stack.append(p)
        else:
            if len(stack)== 0 : return False
            if not stack.pop() == pairs[p]: return False
    
    return len(stack)== 0 # not stack
    
    

def solution(s):
    answer = 0
    parentheses = s
        
    for i in range(len(s)):
        if right_pairs(parentheses): 
            answer += 1
            
        parentheses = parentheses[1:]+parentheses[0]
        
    return answer